src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc

```diff
@@ -86,102 +86,102 @@ class TraversalOrderVisitor
     return Unit();
   }
   Unit PostVisit(const BaseFact& baseFact, Unit result) override {
-    if (post_visits_) nodes_.push_back(baseFact.DebugPrint());
+    if (post_visits_) nodes_.push_back(baseFact);
     return result;
   }
 
   Unit PreVisit(const Fact& fact) override {
-    if (pre_visits_) nodes_.push_back(fact.DebugPrint());
+    if (pre_visits_) nodes_.push_back(fact);
     return Unit();
   }
   Unit PostVisit(const Fact& fact, Unit result) override {
-    if (post_visits_) nodes_.push_back(fact.DebugPrint());
+    if (post_visits_) nodes_.push_back(fact);
     return result;
   }
 
   Unit PreVisit(const ConditionalAssertion& condAssertion) override {
-    if (pre_visits_) nodes_.push_back(condAssertion.DebugPrint());
+    if (pre_visits_) nodes_.push_back(condAssertion);
     return Unit();
   }
   Unit PostVisit(const ConditionalAssertion& condAssertion,
                  Unit result) override {
-    if (post_visits_) nodes_.push_back(condAssertion.DebugPrint());
+    if (post_visits_) nodes_.push_back(condAssertion);
     return result;
   }
 
   Unit PreVisit(const Assertion& assertion) override {
-    if (pre_visits_) nodes_.push_back(assertion.DebugPrint());
+    if (pre_visits_) nodes_.push_back(assertion);
     return Unit();
   }
   Unit PostVisit(const Assertion& assertion, Unit result) override {
-    if (post_visits_) nodes_.push_back(assertion.DebugPrint());
+    if (post_visits_) nodes_.push_back(assertion);
     return result;
   }
 
   Unit PreVisit(const SaysAssertion& saysAssertion) override {
-    if (pre_visits_) nodes_.push_back(saysAssertion.DebugPrint());
+    if (pre_visits_) nodes_.push_back(saysAssertion);
     return Unit();
   }
   Unit PostVisit(const SaysAssertion& saysAssertion, Unit result) override {
-    if (post_visits_) nodes_.push_back(saysAssertion.DebugPrint());
+    if (post_visits_) nodes_.push_back(saysAssertion);
     return result;
   }
 
   Unit PreVisit(const Query& query) override {
-    if (pre_visits_) nodes_.push_back(query.DebugPrint());
+    if (pre_visits_) nodes_.push_back(query);
     return Unit();
   }
   Unit PostVisit(const Query& query, Unit result) override {
-    if (post_visits_) nodes_.push_back(query.DebugPrint());
+    if (post_visits_) nodes_.push_back(query);
     return result;
   }
 
   Unit PreVisit(const Program& program) override {
-    if (pre_visits_) nodes_.push_back(program.DebugPrint());
+    if (pre_visits_) nodes_.push_back(program);
     return Unit();
   }
   Unit PostVisit(const Program& program, Unit result) override {
-    if (post_visits_) nodes_.push_back(program.DebugPrint());
+    if (post_visits_) nodes_.push_back(program);
     return result;
   }
 
-  const std::vector<std::string>& nodes() const { return nodes_; }
+  const std::vector<ASTNode>& nodes() const { return nodes_; }
 
  private:
   bool pre_visits_;
   bool post_visits_;
-  std::vector<std::string> nodes_;
+  std::vector<ASTNode> nodes_;
```

Comment:
This will not do what you want. This is saying "`nodes_` is a `std::vector` containing directly-embedded instances of `ASTNode`". When you look at what an `ASTNode` is, it's underwhelming:

```C++
class ASTNode {
    virtual bool operator==(const ASTNode& otherNode ) const = 0;
};
```

An `ASTNode` contains no members and has no method definitions. It is useless on its own. It is only useful as an interface through which to access other classes.

An annoying thing coming to C++ from languages which emphasize true sum types or which hide their pointers is that the sum-type-ness of using an abstract class in C++ only "activates" if you dereference a pointer. So calling a virtual method on a value of type `ASTNode *` may cause a subclass's method to be called, but calling a virtual method on a value of type `ASTNode` will use the version in `ASTNode` itself to be used only.

This seam comes from both the history and the mechanical implementation of derived classes in C++. It's not far off to say "is a base type of" in C++ is the same as "is a prefix of" the type in its memory layout. If you're writing C in the pre-C++ days, but you find yourself wanting to write derived classes, you end up having some type be the prefix of your family of derived classes, pass around a pointer to that prefix, and put some kind of hint to the full type in that prefix. Then, when you need the full type, you use the hint to perform the cast and widen the pointer's element type to the full type.

C++ builds off of this C pattern by making this cast occur automatically upon the pointer dereference and, rather than forcing you to put your own hint in the prefix, uses the vtable pointer itself as the hint. Giant improvement over the C version!

But, just like in the C version, if you end up copying that prefix, you've essentially "beheaded" the structure. If the interesting stuff was in the tail, you've left it behind, and potentially opened yourself up to a memory error by looking for a tail that isn't there.

To help prevent yourself from making this error, I suggest making the constructor for `ASTNode` `protected`, thus preventing you from accidentally constructing objects of that type directly (as you do here). This should change to be something like `std::vector<const ASTNode *>` or `std::vector<std::unique_ptr<ASTNode>>`, depending on ownership story.